### PR TITLE
fix(pt): fix precision

### PR DESCRIPTION
**Pull request by iProzd**
Tried to implement the decorator as in #4343, but encountered JIT errors.

## Release Notes

- **New Features**
	- Enhanced precision handling across various descriptor classes and methods, ensuring consistent tensor operations.
	- Updated output formats in several classes to improve clarity and usability.
	- Introduced a new environment variable for stricter control over tensor precision handling.
	- Added a new parameter to the `DipoleFittingNet` class for excluding specific types.

- **Bug Fixes**
	- Removed conditions that skipped tests for "float32" data type, allowing all tests to run consistently.

- **Documentation**
	- Improved error messages for dimension mismatches and unsupported parameters, enhancing user understanding.

- **Tests**
	- Adjusted test parameters for consistency in handling `fparam` and `aparam` across multiple test cases.
	- Simplified tensor handling in tests by removing unnecessary type conversions before compression.

### CODE REVIEW

File: deepmd/pt/model/descriptor/dpa1.py [hunk 1]
```diff
@@ -693,7 +697,13 @@ def forward(
...
             g1 = torch.cat([g1, g1_inp], dim=-1)
 
-        return g1, rot_mat, g2, h2, sw
+        return (
+            g1.to(dtype=env.GLOBAL_PT_FLOAT_PRECISION),
+            rot_mat.to(dtype=env.GLOBAL_PT_FLOAT_PRECISION),
+            g2.to(dtype=env.GLOBAL_PT_FLOAT_PRECISION) if g2 is not None else None,
+            h2.to(dtype=env.GLOBAL_PT_FLOAT_PRECISION),
+            sw.to(dtype=env.GLOBAL_PT_FLOAT_PRECISION),
+        )
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/descriptor/dpa1.py:706]: the forward methods of all descriptors has the same interface, then you may want to implement methods like `cast_inputs_types` and `cast_outputs_types` to improve the maintainability of the code.

File: deepmd/pt/model/network/mlp.py [hunk 2]
```diff
@@ -200,7 +200,6 @@ def forward(
             The output.
         """
         ori_prec = xx.dtype
-        xx = xx.to(self.prec)
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/network/mlp.py:203]: add option to remove the type cast in mlp only for debug purpose.

File: deepmd/pt/model/task/fitting.py [hunk 3]
```diff
@@ -495,12 +501,12 @@ def _forward_common(
                     ):
                         atom_property -= ll(xx_zeros)
                 atom_property = atom_property + self.bias_atom_e[type_i]
-                atom_property = atom_property * mask
+                atom_property = torch.where(mask, atom_property, 0.0)
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/fitting.py:504]: I find the change is not necessary.

File: deepmd/pt/model/task/fitting.py [hunk 4]
```diff
@@ -495,12 +501,12 @@ def _forward_common(
...
                 outs = (
                     outs + atom_property
                 )  # Shape is [nframes, natoms[0], net_dim_out]
         # nf x nloc
-        mask = self.emask(atype)
+        mask = self.emask(atype).to(torch.bool)
         # nf x nloc x nod
-        outs = outs * mask[:, :, None]
+        outs = torch.where(mask[:, :, None], outs, 0.0)
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/fitting.py:511]: I find the change is not necessary.

**Review by wanghan-iapcm — COMMENTED** [review]
you may check https://github.com/wanghan-iapcm/deepmd-kit/tree/fix-fit-type for an implementation for fitting.

File: deepmd/pt/model/task/fitting.py (hunk 3 above)
Comment (njzjz) [reply deepmd/pt/model/task/fitting.py:504]: Here `mask` is `bool`, which causes an implicit data type promotion.

**Review by wanghan-iapcm — APPROVED** [review]

**Review by njzjz — APPROVED** [review]
